lib/astarte_data_access/device.ex

```diff
@@ -19,50 +19,55 @@
 defmodule Astarte.DataAccess.Device do
   require Logger
   alias Astarte.Core.Device
-  alias CQEx.Query
-  alias CQEx.Result
+  alias Astarte.DataAccess.Xandra.Utils
 
-  @spec interface_version(:cqerl.client(), Device.device_id(), String.t()) ::
+  @spec interface_version(String.t(), Device.device_id(), String.t()) ::
           {:ok, integer} | {:error, atom}
-  def interface_version(client, device_id, interface) do
-    device_introspection_statement = """
+  def interface_version(realm, device_id, interface_name) do
+    with true <- Utils.verify_realm_ok(realm) do
+      Xandra.Cluster.run(
+        :astarte_data_access_xandra,
+        &do_interface_version(&1, realm, device_id, interface_name)
+      )
+    end
+  end
+
+  defp do_interface_version(conn, realm_name, device_id, interface_name) do
+    device_introspection_query = """
     SELECT introspection
-    FROM devices
+    FROM #{realm_name}.devices
     WHERE device_id=:device_id
     """
 
-    device_introspection_query =
-      Query.new()
-      |> Query.statement(device_introspection_statement)
-      |> Query.put(:device_id, device_id)
+    with {:ok, prepared} <-
+           Xandra.prepare(conn, device_introspection_query),
+         {:ok, %Xandra.Page{} = page} <-
+           Utils.retrieve_page(conn, prepared, %{device_id: device_id}),
+         {:ok, introspection} <- retrieve_introspection(page) do
+      fetch_interface_major(introspection, interface_name)
+    end
+  end
 
-    with {:ok, result} <- Query.call(client, device_introspection_query),
-         device_row when is_list(device_row) <- Result.head(result),
-         introspection <- Keyword.get(device_row, :introspection) || [],
-         {_interface_name, interface_major} <-
-           List.keyfind(introspection, interface, 0, :interface_not_found) do
-      {:ok, interface_major}
-    else
-      :empty_dataset ->
-        Logger.debug("interface_version: device not found #{inspect(device_id)}")
+  defp retrieve_introspection(page) do
+    case Enum.to_list(page) do
+      [] ->
         {:error, :device_not_found}
 
-      :interface_not_found ->
-        # TODO: report device introspection here for debug purposes
-        Logger.warn(
-          "interface_version: interface #{inspect(interface)} not found in device #{inspect(device_id)} introspection"
-        )
-
-        {:error, :interface_not_in_introspection}
+      [%{introspection: introspection}] ->
+        {:ok, introspection}
+    end
+  end
 
-      %{acc: _, msg: error_message} ->
-        Logger.warn("interface_version: database error: #{error_message}")
-        {:error, :database_error}
+  defp fetch_interface_major(introspection, interface_name) do
+    major =
+      Enum.find_value(introspection, fn {name, major} ->
+        if name == interface_name, do: major
+      end)
```

Comment:
Xandra correctly returns maps for map columns

![2023-03-20-160944_1126x315_scrot](https://user-images.githubusercontent.com/2528778/226383078-8b08ce51-c425-4a99-b235-6e0256a7bd50.png)

So just `Map.fetch(introspection, name)`, no need for a dedicated `fetch_interface_major` function